src/ApiCore/GapicClientTrait.php

```diff
@@ -118,56 +119,209 @@ private function setClientOptions(array $options)
+        return $callStack($call, $options);
     }
 
     /**
-     * @return callable
+     * @param Call $call
+     * @param array $options {
+     *     Call Options
+     *
+     *     @type array $headers [optional] key-value array containing headers
+     *     @type int $timeoutMillis [optional] the timeout in milliseconds for the call
+     *     @type array $transportOptions [optional] transport-specific call options
+     * }
+     * @param array $callConstructionOptions {
+     *     Call Construction Options
+     *
+     *     @type RetrySettings $retrySettings [optional] A retry settings override
+     *           For the call.
+     * }
+     * @return ClientStream
      */
-    private function getClientStreamingCallable()
+    private function startClientStreamingCall(Call $call, array $options, array $callConstructionOptions)
     {
-        $callable = function (Call $call, array $options) {
-            return $this->transport->startClientStreamingCall($call, $options);
-        };
-        return new AgentHeaderMiddleware($callable, $this->agentHeaderDescriptor);
+        $callStack = $this->createCallStack(
+            function (Call $call, array $options) {
+                return $this->transport->startClientStreamingCall($call, $options);
+            },
+            $callConstructionOptions
+        );
+        return $callStack($call, $options);
     }
 
     /**
+     * @param Call $call
+     * @param array $options {
+     *     Call Options
+     *
+     *     @type array $headers [optional] key-value array containing headers
+     *     @type int $timeoutMillis [optional] the timeout in milliseconds for the call
+     *     @type array $transportOptions [optional] transport-specific call options
+     * }
+     * @param array $callConstructionOptions {
+     *     Call Construction Options
+     *
+     *     @type RetrySettings $retrySettings [optional] A retry settings override
+     *           For the call.
+     * }
+     * @return ServerStream
+     */
+    private function startServerStreamingCall(Call $call, array $options, array $callConstructionOptions)
+    {
+        return $this->createCallStack(
+            function (Call $call, array $options) {
+                return $this->transport->startServerStreamingCall($call, $options);
+            },
+            $callConstructionOptions
+        );
+    }
+
+    /**
+     * @param callable $callable
+     * @param array $callConstructionOptions {
+     *     Call Construction Options
+     *
+     *     @type RetrySettings $retrySettings [optional] A retry settings override
+     *           For the call.
+     * }
      * @return callable
      */
-    private function getServerStreamingCallable()
+    private function createCallStack(callable $callable, array $callConstructionOptions)
+    {
+        return new RetryMiddleware(
+            new AgentHeaderMiddleware(
+                $callable,
+                $this->agentHeaderDescriptor
```

Comment:
Could this be passed through `$callConstructionOptions` instead of accessed through `$this`?